library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java

```diff
@@ -1337,26 +1340,57 @@ private static boolean formatsMatch(Format manifestFormat, Format sampleFormat)
     return true;
   }
 
-  private static DummyTrackOutput createDummyTrackOutput(int id, int type) {
-    Log.w(TAG, "Unmapped track with id " + id + " of type " + type);
-    return new DummyTrackOutput();
-  }
+  /**
+   * Extend base SampleQueue to add HLS specific processing of the samples, including:
+   *
+   * <ul>
+   *   <li>segment time boundary checks on timestamps of committed samples</li>
+   *   <li>cleaning the {@link Format#metadata} to avoid excessive format changes</li>
+   * </ul>
+   *
+   * The timestamp check verifies that the adjusted sample time (via {@link TimestampAdjuster}) does not
+   * fall outside of a set percentage ({@link #MAX_TIMESTAMP_DEVIATION_PERCENTAGE}) of the time
+   * boundaries of the segment as expressed by the segment duration ((@link HlsMediaChunk#endTimeUs} -
+   * {@link HlsMediaChunk#startTimeUs}).  This is loosely mandated by the Pantos spec and checked by
+   * Apple's mediastreamvalidator.
+   *
+   */
+  private static final class HlsSampleQueue extends SampleQueue {
+
+    private static final String TAG = "HlsSampleQueue";
+
+    /**
+     * largest timestamp deviation from the segment time bounds expressed as a percentage of
+     * the segment duration.
+     */
+    private static double MAX_TIMESTAMP_DEVIATION_PERCENTAGE = 0.50;
+
+    private long lowestTimeUs = C.TIME_UNSET;
+    private long highestTimeUs = C.TIME_UNSET;
 
-  private static final class FormatAdjustingSampleQueue extends SampleQueue {
```

Comment:
Rename `FormatAdjustingSampleQueue` to `HlsSampleQueue` and add checks for timestamp discontinuity.